alacritty/src/renderer/rects.rs

```diff
@@ -190,3 +195,231 @@ impl RenderLines {
+    // TODO these can certainly be i16.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+/// Struct to group together rect-related GL objects and rendering functionality.
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects. VAO stores vertex attributes binding.
+    vao: GLuint,
+    vbo: GLuint,
+
+    program: RectShaderProgram,
+}
+
+impl RectRenderer {
+    /// Update program when doing live-shader-reload.
+    pub fn set_program(&mut self, program: RectShaderProgram) {
+        self.program = program;
+    }
+
+    pub fn new() -> Result<Self, renderer::Error> {
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let program = RectShaderProgram::new()?;
+
+        unsafe {
+            // Allocate buffers.
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+
+            gl::BindVertexArray(vao);
+
+            // VBO binding is not part ot VAO itself, but VBO binding is stored in attributes.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $normalize:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        $normalize,
+                        size_of::<Vertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
+
+            // Position.
+            add_attr!(2, gl::FLOAT, gl::FALSE, f32);
+
+            // Color.
+            add_attr!(4, gl::UNSIGNED_BYTE, gl::TRUE, u8);
+
+            // Reset buffer bindings.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self { vao, vbo, program })
+    }
+
+    pub fn draw(&mut self, size_info: &SizeInfo, rects: Vec<RenderRect>) {
+        unsafe {
+            // Bind VAO to enable vertex attribute slots specified in new().
+            gl::BindVertexArray(self.vao);
+
+            // Bind VBO only once for buffer data upload only.
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(self.program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        // Build rect vertices vector.
+        let mut vertices = RectVertices::new(rects.len());
+        for rect in &rects {
+            vertices.add_rect(center_x, center_y, rect);
+        }
+
+        unsafe {
+            // Upload and render accumulated vertices.
+            vertices.draw();
```

Comment:
I've checked the discussion wrt where to put vector with vertices, and in the current implementation without indices I think it makes sense to store it on a Renderer(since Vertices can't draw themself, renderer can draw something with vertices).

So the change would be simple.

- Store `RectVertices` (or just `Vec`) on the `RectRenderer`
- When starting drawing you call `self.vertices.clear()`.
- Iterate over `rects` and add new rects
- Inline what `draw` does on the `RectVertices`

We'd have rectangles for each frame anyway so allocating vector for 1 element will be just strange.

Doing minimum vsync amount of allocations each frame seems wasteful, especially with slow system allocators.

Reply:
I can't see this 2-3% difference (too much noise here). But I moved allocation back to the parent RectRenderer just in case.